src/main.py

```diff
@@ -0,0 +1,153 @@
+DCT_API_KEY = os.environ.get('DCT_API_KEY')
+NEW_RELIC_INSERT_KEY = os.environ.get('NEW_RELIC_INSERT_KEY')
+
+if (DCT_HOST_URL and DCT_API_KEY and NEW_RELIC_INSERT_KEY) is None:
+    raise Exception("Please add the below keys as environment variables: "
+                    "\n1. DCT_HOST_URL"
+                    "\n2. DCT_API_KEY"
+                    "\n3. NEW_RELIC_INSERT_KEY")
+
+DCT_API_URL = f'https://{DCT_HOST_URL}/v2/'
+QUERY = '?limit=1000'
+DLPX_TYPES = [i.strip() for i in config['COMPONENTS']['monitor'].split(',')]
+
+# Request Headers ...
+req_headers = {
+    'Authorization': DCT_API_KEY  # noqa
+}
+
+# Python session, also handles the cookies ...
+session = requests.session()
+
+
+@tenacity.retry(
+    retry=tenacity.retry_if_exception_type(
+        newrelic_telemetry_sdk.client.HTTPError
+    ),
+    stop=tenacity.stop_after_attempt(5),
+    wait=tenacity.wait_fixed(5),
+    before_sleep=tenacity.before_sleep_log(logging, logging.INFO),
+)
+def send_event(event):
+    logging.debug(event)
+    event_client = EventClient(NEW_RELIC_INSERT_KEY)
+    response = event_client.send(event)
+    response.raise_for_status()
+    logging.debug("Event sent successfully!\n")
+
+
+@tenacity.retry(
+    retry=tenacity.retry_if_exception_type(
+        AssertionError
+    ),
+    stop=tenacity.stop_after_attempt(5),
+    wait=tenacity.wait_fixed(5),
+    before_sleep=tenacity.before_sleep_log(logging, logging.DEBUG),
+)
+def get_data_from_dct(component, query=QUERY):
+    url = f"{DCT_API_URL}{component}{query}"
+    response = requests.get(url, headers=req_headers, verify=False)
+    assert response.status_code == 200
+    response_json = response.json()
+    records = response_json['response_metadata']['total']
+    logging.debug(f"Get {component} returned {records} records")
+
+    # handle pagination
+    if 'next_cursor' in response_json['response_metadata']:
+        logging.debug('Response is paginated getting the next page')
+        next_cursor = response_json['response_metadata']['next_cursor']
+        next_page = get_data_from_dct(
+            component, query=QUERY+'&cursor={}'.format(next_cursor)
+        )
+        response_json['items'].extend(next_page['items'])
+
+    return response_json
+
+
+def push_data():
+    while True:
+        logging.info('*' * 70)
+        logging.info("Pushing Data to New Relic at %s",
+                     datetime.datetime.now())
+        logging.info('*' * 70)
+        for i in DLPX_TYPES:
+            response_json = get_data_from_dct(i)
+
+            new_relic_type = "Delphix " + str(i).split('/')[-1]
+            logging.info("Sending Data for %s \n", new_relic_type)
+            for line in response_json['items']:
+                event = Event(
+                    new_relic_type, line
+                )
+                send_event(event)
+
+        logging.info(f"Sleeping for %s seconds before pushing again", INTERVAL)
+        logging.info('-' * 70)
+        time.sleep(INTERVAL)
```

Comment:
We should add exception handling for communication and network errors here that will catch any momentary communication failures and not exit the whole process. Possible future enhancement.

Reply:
sure arun.

Reply:
I was thinking if we are using a while loop and any failure occurs the program shuts down and does not send anything in future intervals.
For now adding a generic exception inside the for loop where we get data and send it for each component.
If it fails for one component, still it will try for the other components

Also in case of network calls to DCT or New Relic we have tenacity.retry , for that to work we cannot add exceptional handling inside those methods.